Add a class to manage desktop screenshots

mmskv opened the pull request:
Now it only
* Saves preset screen region to specified image format
* Creates new pixmap and copies saved image onto it

Need code review for further development

mmskv committed:
Add DesktopPixmap class

Doesn't have screenshot compressor yet
A lot of FIXMEs and TODOs

mmskv committed:
Add test DesktopPixmap class usage

TheHussky commented on src/DesktopPixmap.cpp:
```diff
@@ -0,0 +1,99 @@
+  this->y = y;
+  this->height = height;
+  this->width = width;
+  this->name = name; // XXX Is this OK?
```
I'm not sure that's ok. Name is a reference itself(why, btw?), a pointer to reference might lead to some unexpected errors. Moreover, the <type & variable> is used in args when it needs to be changed not only within the function, but globally.

mmskv replied:
Looks like it does not matter for short strings.

As I won't change the passed `const` string, only move operations will be called by the compiler. But code generated with `-O3` flag looks the same both for the const reverence and by-value argument passes.

TheHussky commented on src/DesktopPixmap.hpp:
```diff
@@ -0,0 +1,54 @@
+  ~DesktopPixmap ();
+
+  /**
+   * TODO Rename to something more fitting
```
I suggest naming it smth like "save_screen_copy"

mmskv replied:
Too long imo
And `screen` doesn't fit, as this is a different concept in X11

TheHussky commented on src/DesktopPixmap.hpp:
```diff
@@ -0,0 +1,54 @@
+  void putImage (xcb_connection_t *c, xcb_screen_t &screen);
+
+private:
+  // XXX Do I need static here?
```
static methods are used to use class without initializing the object of this class. You can create gc without using this class at all, so not really, I guess. Not quite sure though, tbh.

mmskv replied:
https://stackoverflow.com/a/6446025
Yes, declaring this function as a class member is redundant, but it looks good, so why not

TheHussky commented on src/DesktopPixmap.cpp:
```diff
@@ -0,0 +1,99 @@
+  auto gi_cookie = xcb_get_image (
+      c,            /* Connection */
+      this->format, /* Image format */
+      screen.root,  /* Screenshot relative to root */
```
isn't it screen->root?

TheHussky commented on src/DesktopPixmap.cpp:
```diff
@@ -0,0 +1,99 @@
+DesktopPixmap::~DesktopPixmap ()
+{
+  delete[] this->image_ptr;
+  // TODO Might need to delete pixmap
```
You can set it free via xcb_free_pixmap anytime, I think. If it's not what you mean by delete, then what?

mmskv replied:
Yes, you are right. Will have to free pixmap in the destructor

mmskv replied:
If I do `xcb_disconnect`, will every x object allocated on server be freed?
Sounds reasonable

TheHussky commented on src/DesktopPixmap.cpp:
```diff
@@ -0,0 +1,99 @@
+  uint32_t mask;
+  uint32_t values[2];
+
+  // TODO Figure out correct mask and values
```
I have seen source code for masks somewhere, their names speak for themselves sometimes. I suggest using transparent background by the way.

mmskv replied:
Can you link me the docs for masks and values please

mmskv replied:
https://www.x.org/releases/current/doc/libxcb/tutorial/index.html

TheHussky commented on src/DesktopPixmap.cpp:
```diff
@@ -0,0 +1,99 @@
+{
+  this->pixmap_id = xcb_generate_id (c);
+  xcb_create_pixmap (c, screen.root_depth, this->pixmap_id,
+                     screen.root, // TODO Check if this is OK
```
Are you afraid of segmentation faults? Seemed alright when I tested smth. The screen->root use never caused errors, if I remember correctly. You use &screen, might fail, but I don't think chaces are high.

mmskv replied:
In documentation for pixmaps it says, that `drawable` argument specifies screen, outside of which the pixmap can't be used. LGTM, as pixmaps won't be used on a different screen.

mmskv reviewed (commented):
Have to link this class in CMakeLists.txt

mmskv commented:
Should I create a `xcb_connection_t` for each DesktopPixmap instance?
Passing `*c` everywhere is tedious.
But maybe it takes up too many resources, need to look into it.

TheHussky reviewed (approved):
LGTM. I'll think about a proper name to save_image.

mmskv committed:
Add static Gcontext and Xconnection to class

Minor readability refactors

mmskv committed:
Update example code to use new class methods

Observing a memory leak from xcb_get_image()

mmskv committed:
Fix memory leak by freeing get_image_reply

mmskv committed:
Fix typos. Add a malloc() limitation note

Remove old TODOs
Fix destructor to use new free method
Remove logging from reference implementation (dexpod.cpp)

mmskv committed:
Rename example program for readability

Initialize DesktopPixmap::format in the constructor
Remove dummy code from dexpo.cpp

mmskv committed:
Remove -Wshadow from build flags

Add `example` executable to cmake
Add DesktopPixmap class to cmake build

mmskv committed:
Don't initialize static variables if they exist

And `DesktopPixmap::gc_ = 0` was redundant as static variables are
initialized to zero by default

TheHussky committed:
Nearest-neighbour resize added

mmskv reviewed (commented):
Nice job!
Couple of notes:
* I see you don't have an autoformatter configured. It is essential, because then my autoformatter won't edit your lines of code and mess up commits. So these things need to be changed:
  * Add a whitespace before and after comments (my c++ formatter does this automatically, perhaps you should configure yours too)
  * Separate function definitions with newline (add newline between l126 and l127)
  * Add a newline at the end of the file
  * Use consistent padding around operators. See .clang-format
  * Don't exceed 80 lines of code width. See .clang-tidy
* `char flag` looks ugly imo, i'm sure there is a better way to implement this 

I'll setup a code style pipeline in the meantime

mmskv committed:
Always initialize pixmap in the constructor

Pixmap creation call was located inside an if-case so
pixmap was not created if another DesktopPixmap instance existed.

mmskv committed:
Refactor resize algorithm

mmskv committed:
Change class and struct names to snake_case

I just prefer the way it looks. And xcb library uses snake_case
Also added couple of documentation comments

mmskv committed:
Add xcb drawable class

This class is a base for desktop_pixmap and window classes
Changed PixmapFormat to pixmap_format
